electrum/lnmsg.py

```diff
@@ -348,9 +370,116 @@ def __init__(self, *, for_onion_wire: bool = False):
+                             value: Union[List[Dict[str, Any]], Dict[str, Any]]) -> None:
+        assert fd
+        assert field_type in self.subtypes, f"unknown subtype {field_type}"
+
+        if isinstance(count, int):
+            assert count >= 0, f"{count!r} must be non-neg int"
+        elif count == "...":
+            pass
+        else:
+            raise Exception(f"unexpected field count: {count!r}")
+        if count == 0:
+            return
+
+        if count == 1:
+            assert isinstance(value, dict) or isinstance(value, list)
+            values = [value] if isinstance(value, dict) else value
+        else:
+            assert isinstance(value, list)
+            values = value
+
+        if count == '...':
+            count = len(values)
+        else:
+            assert count == len(values)
+        if count == 0:
+            return
+
+        for record in values:
+            for subtypename, row in self.subtypes[field_type].items():
+                # subtypedata,<subtypename>,<fieldname>,<typename>,[<count>]
+                subtype_field_name = row[2]
+                subtype_field_type = row[3]
+                subtype_field_count_str = row[4]
+
+                subtype_field_count = _resolve_field_count(subtype_field_count_str,
+                                                           vars_dict=record,
+                                                           allow_any=True)
+
+                if subtype_field_name not in record:
+                    raise Exception(f'complex field type {field_type} missing element {subtype_field_name}')
+
+                if subtype_field_type in self.subtypes:
+                    self._write_complex_field(fd=fd,
+                                              field_type=subtype_field_type,
+                                              count=subtype_field_count,
+                                              value=record[subtype_field_name])
+                else:
+                    _write_field(fd=fd,
+                                 field_type=subtype_field_type,
+                                 count=subtype_field_count,
+                                 value=record[subtype_field_name])
+
+    def _read_complex_field(self, *, fd: io.BytesIO, field_type: str, count: Union[int, str])\
+            -> Union[bytes, List[Dict[str, Any]], Dict[str, Any]]:
+        assert fd
+        if isinstance(count, int):
+            assert count >= 0, f"{count!r} must be non-neg int"
+        elif count == "...":
+            pass
+        else:
+            raise Exception(f"unexpected field count: {count!r}")
+        if count == 0:
+            return b""
+
+        parsedlist = []
+
+        # since we don't know the length of the field in advance, and io.BytesIO doesn't provide a look-ahead,
+        # we use this crappy way to detect EOF; read all, and if len(left) > 0, re-wrap in BytesIO stream
+        while left := fd.read(-1):
```

Comment:
Can't we use the existing `_num_remaining_bytes_to_read` function?
Or maybe I don't understand the issue.

Reply:
Not sure why I thought seeking was not possible on `io.BytesIO`..